File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        def reverse(ptr):
            pre = None
            nex = None
            while ptr != None:
                nex = ptr.next
                ptr.next = pre
                pre = ptr
                ptr = nex
            return pre
        if not head or not head.next:
            return True
        fast=curr=head
        while fast.next and fast.next.next:
            fast=fast.next.next
            curr=curr.next
        curr.next=reverse(curr.next)
        curr=curr.next
        dummy=head
        while curr != None:
            if dummy.val != curr.val:
                return False
            dummy = dummy.next
            curr = curr.next
        return True
            
        

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode p1 = head, p2 = reverse(findMid(head));
        
        while ((p2 != null) && (p1.val == p2.val)){
            p1 = p1.next;
            p2 = p2.next;
        }
        return true ? p2 == null : false;
        
    }
    
    private ListNode reverse(ListNode cur){
        ListNode prev = null, temp = null;
        
        while (cur != null){
            temp = cur.next;
            cur.next = prev;
            prev = cur;
            cur = temp;
        }
        return prev;
    }
    
    private ListNode findMid(ListNode cur){
        ListNode slow = cur, fast = cur;
        
        while ((fast != null) && (fast.next != null)){
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }
}